### PR TITLE
fix PhysX pointer loss percision on Android API 30 (#10811)

### PR DESCRIPTION
**Pull request by lealzhan**
Re: #https://github.com/cocos/3d-tasks/issues/12467

### Changelog

* 

-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.

  > Manual trigger with `@cocos-robot run test cases` afterward.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

File: native/cocos/bindings/auto/jsb_physics_auto.cpp [hunk 1]
```diff
@@ -24,24 +24,24 @@ SE_BIND_PROP_GET(js_physics_getter_return_true)
 se::Object* __jsb_cc_physics_RevoluteJoint_proto = nullptr; // NOLINT
 se::Class* __jsb_cc_physics_RevoluteJoint_class = nullptr;  // NOLINT
 
-static bool js_physics_RevoluteJoint_getImpl(se::State& s) // NOLINT(readability-identifier-naming)
+static bool js_physics_RevoluteJoint_getObjectID(se::State& s) // NOLINT(readability-identifier-naming)
```
Comment (dumganhar) [inline native/cocos/bindings/auto/jsb_physics_auto.cpp]: No need to commit files in cocos/bindings/auto/* , they will be generated automatically by cocos robot.

File: native/cocos/physics/physx/PhysXEventManager.cpp [hunk 2]
```diff
@@ -103,8 +103,13 @@ void PhysXEventManager::SimulationEventCallback::onContact(const physx::PxContac
 
 void PhysXEventManager::refreshPairs() {
     for (auto iter = getTriggerPairs().begin(); iter != getTriggerPairs().end();) {
-        const auto &selfIter = getPxShapeMap().find(reinterpret_cast<uintptr_t>(&(reinterpret_cast<PhysXShape *>(iter->get()->shapeA)->getShape())));
-        const auto &otherIter = getPxShapeMap().find(reinterpret_cast<uintptr_t>(&(reinterpret_cast<PhysXShape *>(iter->get()->shapeB)->getShape())));
+        uintptr_t WrapperPtrShapeA = PhysXWorld::getInstance().getWrapperPtrWithObjectID(iter->get()->shapeA);
```
Comment (minggo) [inline native/cocos/physics/physx/PhysXEventManager.cpp]: Please do not capitalize the first letter.

File: native/cocos/physics/physx/PhysXRigidBody.h [hunk 3]
```diff
@@ -80,12 +78,14 @@ class PhysXRigidBody final : public IRigidBody {
...
     void setMask(uint32_t m) override;
     inline uint32_t getInitialGroup() const { return _mGroup; }
+    uint32_t getObjectID() override {return _mObjectID;};
 
 protected:
     // physx::PhysXWorld* mWrappedWorld;
     PhysXSharedBody *_mSharedBody;
     uint32_t _mGroup;
     bool _mEnabled;
+    uint32_t _mObjectID;
```
Comment (minggo) [inline native/cocos/physics/physx/PhysXRigidBody.h]: Please initialize with default values.

File: native/cocos/physics/spec/IBody.h [hunk 4]
```diff
@@ -77,6 +75,7 @@ class IRigidBody : public ILifecycle {
     virtual void setGroup(uint32_t g) = 0;
     virtual uint32_t getMask() = 0;
     virtual void setMask(uint32_t m) = 0;
+    virtual uint32_t  getObjectID() = 0;
```
Comment (minggo) [inline native/cocos/physics/spec/IBody.h]: I think can add const like
```c++
virtual uint32_t  getObjectID() const = 0;
```

File: platforms/native/engine/jsb-physics.js [hunk 5]
```diff
@@ -386,8 +386,8 @@ class Shape {
     onDisable () { this._impl.onDisable(); }
     onDestroy () {
         unBookNode(this._com.node);
-        ptrToObj[this._impl.getImpl()] = null;
-        delete ptrToObj[this._impl.getImpl()];
+        ptrToObj[this._impl.getObjectID()] = null;
+        delete ptrToObj[this._impl.getObjectID()];
```
Comment (minggo) [inline platforms/native/engine/jsb-physics.js:389]: `ptrToObj[this._impl.getObjectID()]` is set to null first, then delete a null pointer?

**Review by dumganhar — APPROVED** [review]

File: native/cocos/physics/sdk/Joint.cpp [hunk 6]
```diff
@@ -57,12 +53,15 @@
         _impl->onDestroy();                         \
     }                                               \
                                                     \
-    void CLASS::setConnectedBody(uintptr_t v) {     \
-        _impl->setConnectedBody(v);                 \
+    void CLASS::setConnectedBody(uint32_t rigidBodyID) {     \
+        _impl->setConnectedBody(rigidBodyID);                 \
```
Comment (minggo) [inline native/cocos/physics/sdk/Joint.cpp]: Please align `\`. 
You can use the tool mentioned [here](https://github.com/cocos/cocos-engine/blob/v3.6.0/docs/CPP_LINTER_AUTOFIX_GUIDE.md) to do it.

File: native/cocos/physics/sdk/Joint.h [hunk 7]
```diff
@@ -39,14 +39,13 @@
...
         ~CLASS() override;                               \
-        uintptr_t getImpl() override;                    \
         void initialize(Node *node) override;            \
         void onEnable() override;                        \
         void onDisable() override;                       \
         void onDestroy() override;                       \
         void setEnableCollision(bool v) override;        \
-        void setConnectedBody(uintptr_t v) override;
-
+        void setConnectedBody(uint32_t rigidBodyID) override;\
```
Comment (minggo) [inline native/cocos/physics/sdk/Joint.h:47]: The same as above.

File: native/cocos/physics/physx/shapes/PhysXTrimesh.cpp [hunk 8]
```diff
@@ -33,9 +33,12 @@ namespace physics {
 
 PhysXTrimesh::PhysXTrimesh() : _mMeshHandle(0),
                                _mConvex(false),
-                               _mIsTrigger(false){};
-
-void PhysXTrimesh::setMesh(uintptr_t handle) {
+                               _mIsTrigger(false),
+                               PhysXShape(){};
```
Comment (minggo) [inline native/cocos/physics/physx/shapes/PhysXTrimesh.cpp]: - don't need `;`
- it is strange to invoke `PhysXShape()` here

File: native/cocos/physics/physx/shapes/PhysXShape.h [hunk 9]
```diff
@@ -79,6 +78,8 @@ class PhysXShape : virtual public IBaseShape {
     physx::PxQuat _mRotation;
     uint8_t _mFlag{0};
     bool _mEnabled{false};
+    uint32_t         _mObjectID{0};
```
Comment (minggo) [inline native/cocos/physics/physx/shapes/PhysXShape.h]: Extra spaces.

File: native/cocos/physics/physx/shapes/PhysXShape.h [hunk 10]
```diff
@@ -71,6 +69,7 @@ class PhysXShape : virtual public IBaseShape {
         return getShape().getFlags().isSet(physx::PxShapeFlag::eTRIGGER_SHAPE);
     }
     void updateFilterData(const physx::PxFilterData &data);
+    uint32_t                getObjectID() override {return _mObjectID;};
```
Comment (minggo) [inline native/cocos/physics/physx/shapes/PhysXShape.h]: Extra spaces.

File: native/cocos/physics/physx/shapes/PhysXCone.cpp [hunk 11]
```diff
@@ -33,9 +33,12 @@
 namespace cc {
 namespace physics {
 
-PhysXCone::PhysXCone() : _mMesh(nullptr){};
+PhysXCone::PhysXCone() : _mMesh(nullptr),
+                         PhysXShape(){};
```
Comment (minggo) [inline native/cocos/physics/physx/shapes/PhysXCone.cpp]: - `;` is not needed
- PhysXShape is base class of `PhysXCone `? Even so, i don't think should invoke the constructor explicitly.

File: native/cocos/physics/physx/PhysXWorld.cpp [hunk 12]
```diff
@@ -315,5 +329,46 @@ RaycastResult &PhysXWorld::raycastClosestResult() {
...
+};
+
+void PhysXWorld::removePXObject(uint32_t PXobjectID) {
+    _mPXObjects.erase(PXobjectID);
+}
+
+uintptr_t PhysXWorld::getPXPtrWithPXObjectID(uint32_t PXObjectID) {
+    if (_mPXObjects.find(PXObjectID) == _mPXObjects.end())
+        return 0;
+    return _mPXObjects[PXObjectID];
```
Comment (minggo) [inline native/cocos/physics/physx/PhysXWorld.cpp]: I think can just cache the iterator for return value like this
```c++
const iter = _mPXObjects.find(PXObjectID);
if (iter == _mPXObjects.end()) {
    return 0;
}
return iter.second;
```

File: native/cocos/physics/physx/PhysXWorld.cpp [hunk 13]
```diff
@@ -315,5 +329,46 @@ RaycastResult &PhysXWorld::raycastClosestResult() {
     return hit;
 }
 
+uint32_t PhysXWorld::addPXObject(uintptr_t PXObjectPtr) {
+    static uint32_t sPXObjectID = 0;
```
Comment (minggo) [inline native/cocos/physics/physx/PhysXWorld.cpp]: We should avoid static variable as possible. As we support `game.restart()` which will restart app in runtime. If using a static variable like this, then hard to restart. Use static member variable is better, then physics module should manage it when restart.

File: native/cocos/physics/physx/PhysXWorld.cpp [hunk 14]
```diff
@@ -193,7 +196,18 @@ uintptr_t PhysXWorld::createMaterial(uint16_t id, float f, float df, float r,
         mat->setFrictionCombineMode(physx::PxCombineMode::Enum(m0));
         mat->setRestitutionCombineMode(physx::PxCombineMode::Enum(m1));
     }
-    return uintptr_t(mat);
+    return true;
+}
+
+uintptr_t PhysXWorld::getPXMaterialPtrWithMaterialID(uint32_t materialID) {
+    physx::PxMaterial *mat;
```
Comment (minggo) [inline native/cocos/physics/physx/PhysXWorld.cpp]: It seems `mat` is useless.

File: native/cocos/bindings/auto/jsb_physics_auto.cpp (hunk 1 above)
Comment (lealzhan) [reply native/cocos/bindings/auto/jsb_physics_auto.cpp]: OK

File: native/cocos/physics/sdk/Joint.cpp (hunk 6 above)
Comment (lealzhan) [reply native/cocos/physics/sdk/Joint.cpp]: OK

File: platforms/native/engine/jsb-physics.js (hunk 5 above)
Comment (lealzhan) [reply platforms/native/engine/jsb-physics.js:389]: It's weired to me, too. Just follow the old code...

Comment (minggo) [reply platforms/native/engine/jsb-physics.js:389]: I think should delete first, then set it to null.

File: native/cocos/physics/physx/PhysXWorld.cpp (hunk 12 above)
Comment (lealzhan) [reply native/cocos/physics/physx/PhysXWorld.cpp]: Seems good

File: native/cocos/bindings/auto/jsb_physics_auto.h [hunk 15]
```diff
@@ -278,8 +277,8 @@ bool js_register_cc_physics_TerrainShape(se::Object *obj); // NOLINT
 SE_DECLARE_FUNC(js_physics_TerrainShape_getAABB);
 SE_DECLARE_FUNC(js_physics_TerrainShape_getBoundingSphere);
 SE_DECLARE_FUNC(js_physics_TerrainShape_getGroup);
-SE_DECLARE_FUNC(js_physics_TerrainShape_getImpl);
```
Comment (minggo) [inline native/cocos/bindings/auto/jsb_physics_auto.h]: Don't have to commit these changes as CI will generate binding codes automatically.

File: native/cocos/physics/physx/PhysXWorld.cpp [hunk 16]
```diff
@@ -124,17 +127,18 @@ void PhysXWorld::setCollisionMatrix(uint32_t index, uint32_t mask) {
...
-uintptr_t PhysXWorld::createConvex(ConvexDesc &desc) {
+uint32_t PhysXWorld::createConvex(ConvexDesc &desc) {
     physx::PxConvexMeshDesc convexDesc;
     convexDesc.points.count = desc.positionLength;
     convexDesc.points.stride = sizeof(physx::PxVec3);
     convexDesc.points.data = static_cast<physx::PxVec3 *>(desc.positions);
     convexDesc.flags = physx::PxConvexFlag::eCOMPUTE_CONVEX;
     physx::PxConvexMesh *convexMesh = getCooking().createConvexMesh(convexDesc, PxGetPhysics().getPhysicsInsertionCallback());
-    return reinterpret_cast<uintptr_t>(convexMesh);
+    uint32_t PXobjectID = addPXObject(reinterpret_cast<uintptr_t>(convexMesh));
```
Comment (SantyWang) [inline native/cocos/physics/physx/PhysXWorld.cpp]: Variable names should start with lowercase

File: native/cocos/physics/physx/PhysXWorld.cpp [hunk 17]
```diff
@@ -149,10 +153,11 @@ uintptr_t PhysXWorld::createTrimesh(TrimeshDesc &desc) {
         meshDesc.triangles.data = static_cast<physx::PxU32 *>(desc.triangles);
     }
     physx::PxTriangleMesh *triangleMesh = getCooking().createTriangleMesh(meshDesc, PxGetPhysics().getPhysicsInsertionCallback());
-    return reinterpret_cast<uintptr_t>(triangleMesh);
+    uint32_t PXobjectID = addPXObject(reinterpret_cast<uintptr_t>(triangleMesh));
```
Comment (SantyWang) [inline native/cocos/physics/physx/PhysXWorld.cpp]: Variable names should start with lowercase

File: native/cocos/physics/physx/PhysXWorld.cpp [hunk 18]
```diff
@@ -171,10 +176,11 @@ uintptr_t PhysXWorld::createHeightField(HeightFieldDesc &desc) {
     hfDesc.samples.stride = sizeof(physx::PxHeightFieldSample);
     physx::PxHeightField *hf = getCooking().createHeightField(hfDesc, PxGetPhysics().getPhysicsInsertionCallback());
     delete[] samples;
-    return reinterpret_cast<uintptr_t>(hf);
+    uint32_t PXobjectID = addPXObject(reinterpret_cast<uintptr_t>(hf));
```
Comment (SantyWang) [inline native/cocos/physics/physx/PhysXWorld.cpp]: Variable names should start with lowercase

**Review by SantyWang — APPROVED** [review]

File: native/cocos/physics/physx/PhysXRigidBody.cpp [hunk 19]
```diff
@@ -37,8 +37,17 @@ using physx::PxVec3;
 namespace cc {
 namespace physics {
 
-PhysXRigidBody::PhysXRigidBody() : _mEnabled(false),
-                                   _mGroup(1) {}
+PhysXRigidBody::PhysXRigidBody() : _mSharedBody(nullptr),
+                                   _mEnabled(false),
+                                   _mGroup(1),
+                                   _mObjectID(0)
```
Comment (minggo) [inline native/cocos/physics/physx/PhysXRigidBody.cpp]: I think it is better to initialize default value in header file when define a class.

File: native/cocos/physics/physx/joints/PhysXDistance.cpp [hunk 20]
```diff
@@ -32,7 +32,8 @@ namespace cc {
 namespace physics {
 
 PhysXDistance::PhysXDistance() : _mPivotA(physx::PxZero),
-                                 _mPivotB(physx::PxZero) {}
+                                 _mPivotB(physx::PxZero),
+                                 PhysXJoint(){}
```
Comment (minggo) [inline native/cocos/physics/physx/joints/PhysXDistance.cpp]: I think don't have to invoke PhysXJoint() explicitly.

File: native/cocos/physics/physx/PhysXRigidBody.cpp (hunk 19 above)
Comment (lealzhan) [reply native/cocos/physics/physx/PhysXRigidBody.cpp]: ok

File: native/cocos/physics/physx/joints/PhysXDistance.cpp (hunk 20 above)
Comment (lealzhan) [reply native/cocos/physics/physx/joints/PhysXDistance.cpp]: ok

**Review by minggo — APPROVED** [review]

**Pull request merged**